### DISCUSSION TITLE
Exception of type: DYNAMIC_LIBRARY_FAILURE .... Error 193 while loading DLL libhsl.dll: %1 is not a valid Win32 application

### DISCUSSION THREAD

**Discussion by brunomorampc** [casadi/casadi#3176] [Q&A]
I am trying to solve an optimization problem with casadi using the HLS linear solvers

```
import casadi as ca

 x = ca.SX.sym('x', 4)
f = x[0] * x[3] * ca.sum1(x[:3]) + x[2]
g = []
g_lb = []
g_ub = []
g.append(ca.sum1(x ** 2))
g_lb.append(40)
g_ub.append(40)

g.append(x[0] * x[1] * x[2] * x[3])
g_lb.append(25)
g_ub.append(ca.inf)

g = ca.vertcat(*g)


# starting point
x0 = ca.DM([1.0, 5.0, 5.0, 1.0])

# variable bounds: 1 <= x[i] <= 5

# executing the solver
nlp = {'x': x, 'f': f, 'g': g}
opts = {'ipopt.linear_solver': 'ma27'}
solver = ca.nlpsol('S', 'ipopt', nlp, opts)
res = solver(
  lbx=[0, 0, 0, 0],
  ubx=[5, 5, 5, 5],
  lbg=g_lb,
  ubg=g_ub,
  x0 = x0
)
print(res)

```
But I get this error ( only if I try to use HSL solvers, otherwise it works)

```
Process finished with exit code 0
Exception of type: DYNAMIC_LIBRARY_FAILURE in file "/work/build/external_projects/src/ipopt-external/src/Common/IpLibraryLoader.cpp" at line 59:
 Exception message: Error 193 while loading DLL libhsl.dll: %1 is not a valid Win32 application
```

I am using Python 3.10 64bit installed with Anaconda on Windows 64bit
In the virtual environment, I installed cyipopt because I want to compare the performances of JAX and CasADi. For that I used the instructions [here ](https://github.com/mechmotum/cyipopt/blob/master/docs/source/install.rst), during this procedure I also installed IPOPT with conda forge as described. I know that casadi comes with IPOPT already and  I suspect I messed something up and now casadi is trying to get the wrong IPOPT? could that be? 

**EDIT**
I installed a new conda environment where I did not install cyipopt and I get a similar error...

```
Exception of type: DYNAMIC_LIBRARY_FAILURE in file "D:\bld\ipopt_1657787518585\work\src\Common\IpLibraryLoader.cpp" at line 59:
 Exception message: Error 193 while loading DLL libhsl.dll: %1 is not a valid Win32 application.
```

**EDIT 2**
I also tried to use virtualenv instead of Anaconda and I get the same error.
Old virtual environments work, so it seems it is something affecting the new virtualenvs I am creating...

**EDIT 3**

In the meanwhile I realized the compiled libraries of the hsl solvers are for win32. This could explain the problem, it is strange though that I did not have this problem so far. I could use the win32 libraries without any problem with windows64...

**Comment by jgillis** [comment]
I suggest to use a dll checker like https://github.com/lucasg/Dependencies/releases to inspect if there are missing dependencies for libhsl.dll . Did you manage to load the dll using ctypes?

**Reply by jgillis** [reply]
https://github.com/casadi/casadi/discussions/3083
